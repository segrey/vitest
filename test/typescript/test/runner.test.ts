import { resolve } from 'pathe'
import fg from 'fast-glob'
import { execa } from 'execa'
import { describe, expect, it } from 'vitest'

describe('should fails', async () => {
  const root = resolve(__dirname, '../failing')
  const files = await fg('*.test-d.*', { cwd: root })

  it('typecheck files', async () => {
    // in Windows child_process is very unstable, we skip testing it
    if (process.platform === 'win32' && process.env.CI)
      return

    const { stderr } = await execa('npx', [
      'vitest',
      'typecheck',
      '--dir',
      'failing',
      '--config',
      resolve(__dirname, './vitest.config.ts'),
    ], {
      cwd: root,
      reject: false,
      env: {
        ...process.env,
        CI: 'true',
        NO_COLOR: 'true',
      },
    })

    expect(stderr).toBeTruthy()
    const lines = String(stderr).split(/\n/g)
    const msg = lines
      .filter(i => i.includes('TypeCheckError: '))
      .reverse()
      .join('\n')
      .trim()
      .replace(root, '<rootDir>')
    expect(msg).toMatchSnapshot()

    files.forEach((file) => {
      const index = lines.findIndex(val => val.includes(`${file}:`))
      const msg = lines.slice(index, index + 8).join('\n')
      expect(msg).toMatchSnapshot(file)
    })
  }, 30_000)
})
