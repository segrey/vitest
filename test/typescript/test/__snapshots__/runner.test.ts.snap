// Vitest Snapshot v1

exports[`should fails > typecheck files > expect-error.test-d.ts 1`] = `
" ❯ expect-error.test-d.ts:4:3
      2| 
      3| test('failing test with expect-error', () => {
      4|   // @ts-expect-error expect nothing
       |   ^
      5|   expectTypeOf(1).toEqualTypeOf<number>()
      6| })
"
`;

exports[`should fails > typecheck files > fail.test-d.ts 1`] = `
" ❯ fail.test-d.ts:4:19
      2| 
      3| test('failing test', () => {
      4|   expectTypeOf(1).toEqualTypeOf<string>()
       |                   ^
      5| })
      6| 
"
`;

exports[`should fails > typecheck files > js-fail.test-d.js 1`] = `
" ❯ js-fail.test-d.js:6:19
      4| 
      5| test('js test fails', () => {
      6|   expectTypeOf(1).toBeArray()
       |                   ^
      7| })
      8| 
"
`;

exports[`should fails > typecheck files > only.test-d.ts 1`] = `
" ❯ only.test-d.ts:4:19
      2| 
      3| test.only('failing test', () => {
      4|   expectTypeOf(1).toEqualTypeOf<string>()
       |                   ^
      5| })
      6| 
"
`;

exports[`should fails > typecheck files 1`] = `
"TypeCheckError: Expected 1 arguments, but got 0.
TypeCheckError: Expected 1 arguments, but got 0.
TypeCheckError: Expected 1 arguments, but got 0.
TypeCheckError: Expected 1 arguments, but got 0.
TypeCheckError: Expected 1 arguments, but got 0.
TypeCheckError: Unused '@ts-expect-error' directive."
`;
